File: cinst/scripts/csv2db.py
#!/usr/bin/env python3
import sqlite3
import glob
import os
import argparse 
#pid = 46051
# parser = argparse.ArgumentParser()
# parser.add_argument('pid', type=int, help='pid of jvm')
# args = parser.parse_args()
# pid = args.pid
pid = os.path.abspath('.').split('-')[-1]
int(pid)
#pid = 58371
dbname = f'{pid}.db'

conn = sqlite3.connect(dbname)
cursor = conn.cursor()
'''
For PutField/AAStore, ADDR0 means holder and ADDR1 means ref;
for GC, ADDR0 means src and ADDR1 means dest;
for NEW, ADDR0 means object address.

For AAStore, ID means array index;
for NEw, ID means object type id;
for PutField, ID means field id.
| 64    | 64    | 64        | 32    | 32    | 32            | 8             | -             |
| TIME  | ADDR0 | ADDR1     | ID    | LINE  | CLASS_NAME_ID | TYPE          | TID           |
| time  | addr  | obj_size  | TYPEID| LINE  | CLASS_NAME_ID | n(new)        | primary key   |
| time  | holder| ref       | INDEX | LINE  | CLASS_NAME_ID | a(aastore)    | ~             |
| time  | holder| ref       | FIELD | LINE  | CLASS_NAME_ID | p(putfield)   | ~             |
| time  | src   | dest      | NULL  | NULL  | NULL          | g(gc)         | ~             |
| time  | addr  | NULL      | NULL  | LINE  | CLASS_NAME_ID | u(use)        | ~             |
| time  | NULL  | NULL      | EID   | NULL  | NULL          | e(event)      | ~             |
| time  | holder| ref       | cid   | LINE  | CLASS_NAME_ID | c(container)  | ~             | cop       | 
'''
# cursor.execute('''--sql
# CREATE TABLE IF NOT EXISTS TIMELINE(
#     TIME BIGIT,
#     ADDR0 BIGINT,
#     ADDR1 BININT,
#     ID INT,
#     LINE INT,
#     CLASS_NAME_ID INT,
#     TYPE CHAR,
#     TID INTEGER PRIMARY KEY AUTOINCREMENT
#     );''')
def get_files(prefix, pid):
    return glob.glob(f'{prefix}-*-{pid}')

class DataType:
    def __init__(self, values, typech, table):
        self.values = values
        self.typech = ord(typech)
        self.table = table
datatype_map = {
    'gc': DataType('TIME,ADDR0,ADDR1', 'g', 'TIMELINE'),
    'AAStore': DataType('LINE, CLASS_NAME_ID, TIME, ID, ADDR0, ADDR1', 'a', 'TIMELINE'),
    'PutField': DataType('LINE, CLASS_NAME_ID, ID, TIME, ADDR0, ADDR1', 'p', 'TIMELINE'),
    "NEW": DataType('LINE, CLASS_NAME_ID, TIME, ID, ADDR0', 'n', 'TIMELINE'),
}

info_map = {
    'classname': DataType('CLASSNAME', 'c', 'CLASSNAME'),
    'typemap': DataType('ID, TYPE', 't', 'TYPEMAP'),
    'field': DataType('ID, FIELD', 'f', 'FIELD')
}

#for datatype in datatype_map:
#    #print(datatype)
#    files = get_files(datatype, pid)
#    for file in files:
#        with open(file) as f:
#            data = f.readlines()
#        cursor.execute('BEGIN TRANSACTION;')
#        for i, line in enumerate(data):
#            #print(f'INSERT INTO TIMELINE ({datatype_map[datatype].values},TYPE) VALUES({line.strip()}, {datatype_map[datatype].typech})')
#            if line.count(',') != datatype_map[datatype].values.count(','):
#                print(f'[WARNING] Line is broken at {file} line {i} with {line}', flush=True)
#            else:
#                cursor.execute(f'INSERT INTO TIMELINE ({datatype_map[datatype].values},TYPE) VALUES({line.strip()}, {datatype_map[datatype].typech})')
#        cursor.execute('COMMIT;')
# cursor.execute('CREATE INDEX IF NOT EXISTS TIME_INDEX ON TIMELINE(TIME)')
cursor.execute('''CREATE TABLE IF NOT EXISTS CLASSNAME(
    ID INTEGER PRIMARY KEY AUTOINCREMENT,
    CLASSNAME TEXT
);
               ''')
cursor.execute('''CREATE TABLE IF NOT EXISTS TYPEMAP(
    ID INT ,
    TYPE TEXT
);
               ''')
cursor.execute('''CREATE TABLE IF NOT EXISTS FIELD(
    ID INT ,
    FIELD TEXT
);
''')
for info in info_map:
    file_list = get_files(info,pid)
    if len(file_list) == 0:
        continue
    file = get_files(info,pid)[0]
    with open(file) as f:
        data = f.readlines()
        holder = ','.join(['?']*len(info_map[info].values.split(',')))
        print(file)
        for line in data:
            #print(f'INSERT INTO {info_map[info].table} ({holder}) VALUES({line})', *line.split(','))
            #cursor.execute(f'INSERT INTO {info_map[info].table} ({info_map[info].values}) VALUES({line})')
            cursor.execute(f'INSERT INTO {info_map[info].table} ({info_map[info].values}) VALUES({holder})', line.strip().split(','))
##file = get_files('typemap',pid)[0]
##with open(file) as f:
##    data = f.readlines()
##    for line in data:
##        cursor.execute(f'INSERT INTO TYPEMAP ({info_map[datatype].values}) VALUES({line})')
cursor.close()
conn.commit()
conn.close()

File: cinst/scripts/src/csv2db.cpp
#include "sqlite3.h"
#include "recorder.hpp"
#include <stdio.h>
#include <string.h>
#include <string>
#include <stdlib.h>
#include <vector>
#include <memory>
#include <unordered_map>
#include <filesystem>
#include "fs_util.h"
//#include <filesystem>
using namespace std;

constexpr size_t BUF_SIZE = (16L<<30);

#define SQLITE3_EXEC( db, sql, call_back, call_back_args, err_msg)\
{ \
    int rc = sqlite3_exec(db, sql, call_back, call_back_args, err_msg); \
    if (rc != SQLITE_OK) { \
        fprintf(stderr, "Sqlite error at %s:%d with %s", __FILE__, __LINE__, *err_msg); \
        sqlite3_free(*err_msg); \
        exit(-1); \
    }\
} 

//void glob(string path, vector<string>& file, string ext)
//{
//    struct stat s;
//    lstat(path.c_str(), &s);
//    if (!S_ISDIR(s.st_mode)) {
//        fprintf(stderr, "%s is not dir\n", path.c_str());
//        exit(-1);
//    }
//
//    struct dirent* filename;
//    DIR * dir;
//    dir = opendir(path.c_str());
//    if (NULL == dir) {
//        fprintf(stderr, "can not open dir %s\n", path.c_str());
//        exit(-1);
//    }
//
//    while ((filename = readdir(dir)) != NULL) 
//    {
//        if( strcmp( filename->d_name , "." ) == 0 || 
//			strcmp( filename->d_name , "..") == 0    )
//			continue;
//        string namestr = string(filename);
//        if (namestr.length() <= )
//    }
//}
//

void glob(const string& pathstr, vector<string>& files, const string& ext) {
    filesystem::path path(pathstr);
    if (!filesystem::exists(path) || !filesystem::is_directory(path)) {
        fprintf(stderr, "dir %s not exist\n", pathstr.c_str());
    }

    filesystem::directory_iterator dir(path);
    for (auto& ite: dir) {
        if (ite.status().type() == filesystem::file_type::regular) {
            string filename = ite.path().filename().string();
            if (endsWith(filename, ext)) {
                files.push_back(filename);
            }
        }
    }
}

const vector<string> prefix_list = {
    "gc", "AAStore", "PutField", "NEW"
};
unordered_map<string, string> sql_map = {
    {"gc", "TIME,ADDR0,ADDR1"},
    {"AAStore", "LINE, CLASS_NAME_ID, TIME, ID, ADDR0, ADDR1"},
    {"PutField", "LINE, CLASS_NAME_ID, ID, TIME, ADDR0, ADDR1"},
    {"NEW", "LINE, CLASS_NAME_ID, TIME, ID, ADDR0, ADDR1"},
    {"Use", "LINE, CLASS_NAME_ID, ADDR0, TIME"},
    {"Free", "TIME, ADDR0, ADDR1"},
    {"Event", "TIME, ID"}
};

unordered_map<string, string> value_map = {
    {"gc", "?,?,?,?"},
    {"AAStore", "?,?,?,?,?,?,?"},
    {"PutField", "?,?,?,?,?,?,?"},
    {"NEW", "?,?,?,?,?,?,?"},
    {"Use", "?,?,?,?,?"},
    {"Free", "?,?,?,?"},
    {"Event", "?,?,?"}
};
unordered_map<string, char> type_map = {
    {"gc",'g'},
    {"AAStore",'a'},
    {"PutField",'p'},
    {"NEW",'n'},
    {"Use", 'u'},
    {"Free", 'f'},
    {"Event", 'e'}
};

void sql_bind(sqlite3_stmt* pstmt, EventRecord<0> *r) {
    int nCol = 1;
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int(pstmt, nCol++, r->event_id);
    sqlite3_bind_int(pstmt, nCol++, 'e');
}
void sql_bind(sqlite3_stmt* pstmt, FreeRecord<0> *r) {
    int nCol = 1;
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int(pstmt, nCol++, r->addr);
    sqlite3_bind_int(pstmt, nCol++, r->obj_size);
    sqlite3_bind_int(pstmt, nCol++, 'f');
}
void sql_bind(sqlite3_stmt* pstmt, UseRecord<0> *r) {
    int nCol = 1;
    sqlite3_bind_int(pstmt, nCol++, r->line);
    sqlite3_bind_int(pstmt, nCol++, r->class_name_id);
    sqlite3_bind_int64(pstmt, nCol++, r->addr);
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int(pstmt, nCol++, 'u');
}
void sql_bind(sqlite3_stmt* pstmt, GCRecord<0>* r) {
    int nCol = 1;
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int64(pstmt, nCol++, r->org_addr);
    sqlite3_bind_int64(pstmt, nCol++, r->new_addr);
    sqlite3_bind_int(pstmt, nCol++, 'g');
}

void sql_bind(sqlite3_stmt* pstmt, AAStoreRecord<0>* r) {
    int nCol = 1;
    sqlite3_bind_int(pstmt, nCol++, r->line);
    sqlite3_bind_int(pstmt, nCol++, r->class_name_id);
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int64(pstmt, nCol++, r->index);
    sqlite3_bind_int64(pstmt, nCol++, r->holder_addr);
    sqlite3_bind_int64(pstmt, nCol++, r->ref_addr);
    sqlite3_bind_int(pstmt, nCol++, 'a');
}
void sql_bind(sqlite3_stmt* pstmt, NewRecord<0>* r) {
    int nCol = 1;
    sqlite3_bind_int(pstmt, nCol++, r->line);
    sqlite3_bind_int(pstmt, nCol++, r->class_name_id);
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int64(pstmt, nCol++, r->obj_typeid);
    sqlite3_bind_int64(pstmt, nCol++, r->addr);
    sqlite3_bind_int64(pstmt, nCol++, r->obj_size);
    sqlite3_bind_int(pstmt, nCol++, 'n');
}
void sql_bind(sqlite3_stmt* pstmt, PutFieldRecord<0>* r) {
    int nCol = 1;
    sqlite3_bind_int(pstmt, nCol++, r->line);
    sqlite3_bind_int(pstmt, nCol++, r->class_name_id);
    sqlite3_bind_int(pstmt, nCol++, r->field_id);
    sqlite3_bind_int64(pstmt, nCol++, r->time);
    sqlite3_bind_int64(pstmt, nCol++, r->holder_addr);
    sqlite3_bind_int64(pstmt, nCol++, r->ref_addr);
    sqlite3_bind_int(pstmt, nCol++, 'p');
}

template<typename Record>
void load_timeline(sqlite3* db, char* _, stringstream& ss,  FILE* fin, char* buf) {
    char* err_msg = NULL;
    char* pzTail;
    static sqlite3_stmt *pstmt = nullptr;
    if (pstmt == nullptr) {
        string ssql = string("INSERT INTO TIMELINE(")+sql_map[Record::prefix]+",TYPE) VALUES("+value_map[Record::prefix]+")";
        const char* sql = ssql.c_str();
        sqlite3_prepare_v2(db, sql, strlen(sql), &pstmt, NULL);
    }

    while (true) {
        size_t size = fread(buf, sizeof(Record), BUF_SIZE/sizeof(Record), fin);
        if (size <= 0) {
            return;
        }
        for (size_t i = 0; i < size; ++i) {
            //ss.str("");
            //ss<<"INSERT INTO TIMELINE ("<<sql_map[string(Record::prefix)]<<",TYPE"<<") VALUES("<<((Record*)buf)[i]<<","<<(int)type_map[Record::prefix]<<")";
            //printf("%s\n", ss.str().c_str()); 
            //SQLITE3_EXEC(db, ss.str().c_str(), NULL, NULL, &err_msg);
            sqlite3_reset(pstmt);
            sql_bind(pstmt, ((Record*)buf)+i);
            sqlite3_step(pstmt);
        }
    }
}

//template <>
//void load_timeline<GCRecord>(sqlite3* db, char* sql, stringstream& ss,  FILE* fin, char* buf)
//{
//    char* err_msg = NULL;
//    char* pzTail;
//    sqlite3_stmt *pstmt;
//    auto sql = "INSERT INTO TIMELINE("+sql_map[string(GCRecord::prefix)]+",TYPE) VALUES(?,?,?,?)";
//    sqlite3_prepare_v2(db, sql, strlen(sql), &pstmt, &pzTail);
//    while (true) {
//        size_t size = fread(buf, sizeof(GCRecord), BUF_SIZE/sizeof(GCRecord), fin);
//        if (size <= 0) {
//            return;
//        }
//        for (size_t i = 0; i < size; ++i) {
//            int nCol = 1;
//            GCRecord* r = ((GCRecord*)buf)+i;
//            //ss.str("");
//            //ss<<"INSERT INTO TIMELINE ("<<sql_map[string(GCRecord::prefix)]<<",TYPE"<<") VALUES("<<((Record*)buf)[i]<<","<<(int)type_map[Record::prefix]<<")";
//            ////printf("%s\n", ss.str().c_str()); 
//            //SQLITE3_EXEC(db, ss.str().c_str(), NULL, NULL, &err_msg);
//        }
//    }
//
//}

//template<typename... Ts>
//void call_load_timeline(const string& filename, Ts&... args) {
//}

#define LOAD_TIMELINE(type, filename, ...) \
if (startsWith(filename, type::prefix)) { \
    load_timeline<type>(__VA_ARGS__); \
}

template<typename T, typename... Ts>
void call_load_timeline(const string& filename, Ts && ... args) {
    if (startsWith(filename, T::prefix)) {
        load_timeline<T>(filename, std::forward<Ts>(args)...);
    }
}
//template<typename T, typename... T1, typename... Ts>
//void call_load_timeline(const string& filename, Ts && ... args) {
//    call_load_timeline<T, Ts...>(filename, std::forward<Ts>(args)...);
//    call_load_timeline<T1..., Ts...>(filename, std::forward<Ts>(args)...);
//}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "csv2db <pid>\n");
        exit(-1);
    }
    sqlite3 *db = NULL;
    string pid = string(argv[1]);
    string db_name = pid + ".db";

    char* err_msg = NULL;
    int rc = sqlite3_open(db_name.c_str(), &db);
    //int rc = sqlite3_open(":memory:", &db);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Failed to open db %s: %s\n", db_name.c_str(), sqlite3_errmsg(db));
        sqlite3_close(db);
        exit(-1);
    }
    sqlite3_exec(db, "PRAGMA synchronous = OFF", 0, 0, 0);

    const char * create_timeline_sql =  \
    "CREATE TABLE IF NOT EXISTS TIMELINE("
        "TIME BIGIT,"
        "ADDR0 BIGINT,"
        "ADDR1 BININT,"
        "ID INT,"
        "LINE INT,"
        "CLASS_NAME_ID INT,"
        "TYPE CHAR,"
        "TID INTEGER PRIMARY KEY AUTOINCREMENT"
    ");";

    SQLITE3_EXEC(db, create_timeline_sql, NULL, NULL, &err_msg);
    SQLITE3_EXEC(db, "BEGIN", NULL, NULL, &err_msg);
    char* buf = new char[BUF_SIZE];
    char* sql = new char[4096];
    vector<string> files;
    glob(".",files, pid);
    
    for (auto& file: files) {
        FILE* fin = fopen(file.c_str(), "rb");
        stringstream ss;
        //if (startsWith(file, "gc")) {
        //    load_timeline<GCRecord>(db, sql, ss, fin, buf);
        //}
        //else if (startsWith(file, "AAStore")) {
        //    load_timeline<AAStoreRecord>(db, sql, ss, fin, buf);
        //}

        //call_load_timeline<GCRecord>(file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(GCRecord<0>, file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(AAStoreRecord<0>, file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(PutFieldRecord<0>, file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(NewRecord<0>, file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(UseRecord<0>, file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(FreeRecord<0>, file, db, sql, ss, fin, buf);
        LOAD_TIMELINE(EventRecord<0>, file, db, sql, ss, fin, buf);

        printf("finish %s\n", file.c_str());
    }

    SQLITE3_EXEC(db, "COMMIT", NULL, NULL, &err_msg);
    
    sqlite3_close(db);

}
